### ISSUE TITLE
circular references in type-helpers.d.ts using tsc 3.4rc

### ISSUE THREAD

**Issue by harrybin** [piotrwitek/typesafe-actions#130]
Using tsc 3.4rc a circular reference is detected in type-helpers.d.ts.

tanspiler error message:
_node_modules/typesafe-actions/dist/type-helpers.d.ts
(1,21): Type alias 'ActionType' circularly references itself.

**Comment by harrybin** [comment]
oh, when having a closer look at the code, it might be bug in the new tsc compiler:
[https://github.com/Microsoft/TypeScript/issues/30483](https://github.com/Microsoft/TypeScript/issues/30483)


```ts
export declare type ActionType<ActionCreatorOrMap> = ActionCreatorOrMap extends ActionCreator<StringType> ? ReturnType<ActionCreatorOrMap> : ActionCreatorOrMap extends object ? ActionCreatorMap<ActionCreatorOrMap>[keyof ActionCreatorOrMap] : never;
//...
export declare type ActionCreatorMap<T> = {
    [K in keyof T]: ActionType<T[K]>;
};
```
...seem not to be a circular reference, right?

**Comment by harrybin** [comment]
suggested solution:
```ts
export declare type ActionType<ActionCreatorOrMap> = ActionCreatorOrMap extends ActionCreator<StringType> ? ReturnType<ActionCreatorOrMap> : ActionCreatorOrMap extends object ? ActionCreators<ActionCreatorOrMap[keyof ActionCreatorOrMap]> : never;

type ActionCreators<T> = T extends any ? ActionType<T> : never;
```

**Comment by AustinBrunkhorst** [comment]
Were you able to get the proposed solution to work? I still get the circular reference error with typescript@3.2.4

**Comment by fightingcat** [comment]
Sorry for the incorrect code, I recently only use my phone viewing github. This should be working:
```typescript
export declare type ActionType<ActionCreatorOrMap> = ActionCreatorOrMap extends ActionCreator<StringType> ? ReturnType<ActionCreatorOrMap> : {
    1: ActionCreators<ActionCreatorOrMap[keyof ActionCreatorOrMap]>;
    0: never;
}[ActionCreatorOrMap extends object ? 1 : 0];

type ActionCreators<T> = T extends any ? ActionType<T> : never;
```

**Comment by AustinBrunkhorst** [comment]
@fightingcat that solution is brilliant! I forgot objects were fair game at compile time. TypeScript is pretty neat 👍. Thanks for sharing

**Comment by ethanneff** [comment]
Typescript 3.4.1 just released. Confirming `typesafe-actions` breaks with this update.

```
 FAIL  src/components/Markdown/__tests__/index.tsx
  ● Test suite failed to run

    TypeScript diagnostics (customize using `[jest-config].globals.ts-jest.diagnostics` option):
    src/screens/Debug/screens/Checklists/models/Item/index.tsx:21:27 - error TS2315: Type 'ActionType' is not generic.

     21 export type ItemActions = ActionType<
                                  ~~~~~~~~~~~
     22   | typeof createItem
        ~~~~~~~~~~~~~~~~~~~~~
    ... 
     25   | typeof toggleActiveItem
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~
     26 >;
        ~
```

**Comment by pwentz** [comment]
I'm also running into the same issue @ethanneff is seeing. Looks like TypeScript `3.4.1` breaks `ActionType`. Looks like the new TS version no longer allows `types` to circularly reference themselves:
```
type ActionType<ActionCreatorOrMap> = any
Type alias 'ActionType' circularly references itself.ts(2456)
```

**Comment by piotrwitek** [comment]
@fightingcat thanks for the suggested solution it looks good 👍 
I have published a new version `typesafe-actions@3.3.0-rc` containing @fightingcat fix.
Updated the codesandbox project and all is working great.

**Comment by piotrwitek** [comment]
Related TS issue: https://github.com/Microsoft/TypeScript/issues/29511
